### PR TITLE
To support traceContext in Webpa GET, SET requests

### PR DESCRIPTION
**Pull request by Thanusha-D**

### CODE REVIEW

File: source/app/libpd.c [hunk 1]
```diff
@@ -173,7 +174,22 @@ static void parodus_receive()
                     {
 						memset(res_wrp_msg, 0, sizeof(wrp_msg_t));
                         getCurrentTime(startPtr);
-                        processRequest((char *)wrp_msg->u.req.payload, wrp_msg->u.req.transaction_uuid, ((char **)(&(res_wrp_msg->u.req.payload))));
+			headers_t *res_headers;
+                        res_headers = (headers_t *)malloc(sizeof(headers_t));
+			if(res_headers != NULL) {
```
Comment (shilpa24balaji) [inline source/app/libpd.c:179]: if null, error logging and return needs to be done

File: source/app/libpd.c [hunk 2]
```diff
@@ -173,7 +174,22 @@ static void parodus_receive()
...
+                        	processRequest((char *)wrp_msg->u.req.payload, wrp_msg->u.req.transaction_uuid, ((char **)(&(res_wrp_msg->u.req.payload))), wrp_msg->u.req.headers, res_headers);
+				if(res_headers != NULL && res_headers->headers[0] != NULL && res_headers->headers[1] != NULL) {
+                                	if(strlen(res_headers->headers[0]) > 0 && strlen(res_headers->headers[1]) > 0) {
+                                          	res_headers->count = 2;
```
Comment (shilpa24balaji) [inline source/app/libpd.c:183]: instead of hardcode better to find the size of the string array

File: source/app/libpd.c [hunk 3]
```diff
@@ -173,7 +174,22 @@ static void parodus_receive()
...
+                                          	res_wrp_msg->u.req.headers = res_headers;
+					  	clearTraceContext();
+					  	WalInfo("The response wrp_msg header fields\n");
+                                          	WalInfo("res header count in res_wrp_msg - %d\n", res_wrp_msg->u.req.headers->count);
+                                          	WalInfo("res header 0 in res_wrp_msg - %s\n", res_wrp_msg->u.req.headers->headers[0]);
+                                          	WalInfo("res header 1 in res_wrp_msg - %s\n", res_wrp_msg->u.req.headers->headers[1]);
```
Comment (shilpa24balaji) [inline source/app/libpd.c:189]: Better to find array size and print all elements

File: source/broadband/webpa_rbus.c [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+#include <wdmp-c.h>
+#include <cimplog.h>
+#include "webpa_rbus.h"
+
+static rbusHandle_t rbus_handle;
+static bool isRbus = false;
+
+bool isRbusEnabled()
+{
+        if(RBUS_ENABLED == rbus_checkStatus())
```
Comment (shilpa24balaji) [inline source/broadband/webpa_rbus.c:14]: Where is this rbus_checkStatus defined?

File: source/app/main.c [hunk 5]
```diff
@@ -55,6 +56,11 @@ int main()
 	WalInfo("B4 msgBusInit\n");
 	msgBusInit(pComponentName);
 	WalInfo("After msgBusInit\n");
+	if(isRbusEnabled())
+	{
+	        WalInfo("webpaRbusInit\n");
```
Comment (shilpa24balaji) [inline source/app/main.c:61]: This log is not needed

File: source/broadband/webpa_rbus.c [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+{
+        rbusError_t ret = RBUS_ERROR_BUS_ERROR;
+        if(isRbusInitialized)
+        {
+                if(traceContext[0] != NULL && traceContext[1] != NULL) {
+                       if(strlen(traceContext[0]) > 0 && strlen(traceContext[1]) > 0) {
+			    WalInfo("Invoked setTraceContext function with value traceParent - %s, traceState - %s\n", traceContext[0], traceContext[1]);    
+                            ret = rbusHandle_SetTraceContextFromString(rbus_handle, traceContext[0], traceContext[1]);
+                            if(ret == RBUS_ERROR_SUCCESS) {
+                                  WalInfo("SetTraceContext request success\n");
```
Comment (shilpa24balaji) [inline source/broadband/webpa_rbus.c:61]: Keep this debug level

File: source/broadband/webpa_rbus.c [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        else {
+                WalError("Rbus not initialzed in setTraceContext function\n");
+        }	
+        return ret;
+}
+
+rbusError_t getTraceContext(char* traceContext[])
+{
+        rbusError_t ret = RBUS_ERROR_BUS_ERROR;
+        char traceParent[512] = {'\0'};
```
Comment (shilpa24balaji) [inline source/broadband/webpa_rbus.c:84]: Please check the length of 512 is same as used in rbus. Need to confirm on the length

File: source/broadband/webpa_rbus.c [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+		       else {
+			       WalPrint("traceParent & traceState are empty\n");
+		       }	       
+	      }
+	      else {
+		      WalError("GetTraceContext request failed with error code - %d\n", ret);
+	      }	      
+	}
+        else { 
+              WalInfo("Rbus not initialzed in getTraceContext function\n");
```
Comment (shilpa24balaji) [inline source/broadband/webpa_rbus.c:105]: Change to error

File: source/broadband/webpa_rbus.c [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+{
+        rbusError_t ret = RBUS_ERROR_BUS_ERROR;
+        char traceParent[512] = {'\0'};
+        char traceState[512] = {'\0'};
+	if(isRbusInitialized)
+        {
+	      ret =  rbusHandle_GetTraceContextAsString(rbus_handle, traceParent, sizeof(traceParent), traceState, sizeof(traceState));
+	      if( ret == RBUS_ERROR_SUCCESS) {
+		      if(strlen(traceParent) > 0 && strlen(traceState) > 0) {
+			      WalInfo("GetTraceContext request success\n");
```
Comment (shilpa24balaji) [inline source/broadband/webpa_rbus.c:91]: change to debug

File: source/app/libpd.c (hunk 2 above)
Comment (shilpa24balaji) [reply source/app/libpd.c:183]: we could ignore for now, for future it would be good if dynamic

File: source/include/webpa_adapter.h [hunk 10]
```diff
@@ -14,6 +14,7 @@
 #include <stdint.h>
 #include <wdmp-c.h>
 #include <cimplog.h>
+#include <libparodus.h>
```
Comment (shilpa24balaji) [inline source/include/webpa_adapter.h:17]: What is this for?

File: CMakeLists.txt [hunk 11]
```diff
@@ -46,6 +46,8 @@ include_directories(${INCLUDE_DIR}
                     ${INCLUDE_DIR}/cimplog
                     ${LIBRARY_DIR}/dbus-1.0/include
                     ${INCLUDE_DIR}/dbus-1.0
+		    ${INCLUDE_DIR}/rbus
```
Comment (shilpa24balaji) [inline CMakeLists.txt:49]: Rbus ExternalProject_Add should also be done in cmakelists.txt

**Review by shilpa24balaji — CHANGES_REQUESTED** [review]

File: source/broadband/webpa_rbus.c (hunk 7 above)
Comment (Thanusha-D) [reply source/broadband/webpa_rbus.c:84]: yes, in rbus samples they used 512 only the size for traceParent & traceState

File: source/include/webpa_adapter.h (hunk 10 above)
Comment (Thanusha-D) [reply source/include/webpa_adapter.h:17]: Included this "libparodus.h" header in webpa_adapter.c file, to use headers_t struct from wrp

File: source/broadband/webpa_rbus.c (hunk 4 above)
Comment (Thanusha-D) [reply source/broadband/webpa_rbus.c:14]: This is defined in rbus code

File: source/include/webpa_adapter.h (hunk 10 above)
Comment (sadhyama) [reply source/include/webpa_adapter.h:17]: yes, agree .

File: source/app/libpd.c (hunk 1 above)
Comment (Thanusha-D) [reply source/app/libpd.c:179]: added the changes

File: source/app/main.c (hunk 5 above)
Comment (Thanusha-D) [reply source/app/main.c]: removed

File: source/broadband/webpa_rbus.c (hunk 6 above)
Comment (Thanusha-D) [reply source/broadband/webpa_rbus.c]: ok done

File: source/broadband/webpa_rbus.c (hunk 8 above)
Comment (Thanusha-D) [reply source/broadband/webpa_rbus.c]: ok done

File: CMakeLists.txt (hunk 11 above)
Comment (Thanusha-D) [reply CMakeLists.txt:49]: added the changes

File: source/app/libpd.c [hunk 12]
```diff
@@ -173,7 +174,22 @@ static void parodus_receive()
                     {
 						memset(res_wrp_msg, 0, sizeof(wrp_msg_t));
                         getCurrentTime(startPtr);
-                        processRequest((char *)wrp_msg->u.req.payload, wrp_msg->u.req.transaction_uuid, ((char **)(&(res_wrp_msg->u.req.payload))));
+			headers_t *res_headers;
```
[the remark this answers is not in the document]
Comment (Thanusha-D) [reply source/app/libpd.c]: done

File: source/include/webpa_adapter.h (hunk 10 above)
Comment (shilpa24balaji) [reply source/include/webpa_adapter.h:17]: can we not use wrp header?

**Pull request merged**